File: QuesFromTest/01/sort-boarding-passes.cpp
/* Author: Jitendra Kumar (jitendra.iitp@gmail.com) */
#include<bits/stdc++.h>
using namespace std;
#define endl '\n'
#define MOD 1000000007
#define DEBUG(x) cerr << #x << " is " << (x) << endl;

void sortTheBoardingPasses(unordered_map<string, string> boardingPasses) {
    /*
     * reverse map will contains the mapping in reverse.
     * So for (travel_from_city ---> travel_to_city) in boardingPasses,
     * We will have a reverse mapping, (travel_to_city ---> travel_from_city) entry in reverse map.
     */
    unordered_map<string, string> reverse;
    unordered_map<string, string>::iterator it;
    for(it = boardingPasses.begin(); it != boardingPasses.end(); it++) {
        /*
         * reversing the initial mapping.
         */
        reverse[it->second] = it->first;
    }
    string sourceCityName = "";
    for(it = boardingPasses.begin(); it != boardingPasses.end(); it++) {
        if(reverse.find(it->first) == reverse.end()) {
            /*
             * The city which does not have a 'travel_from_city' entry in reverse map,
             * Actually is the starting point because you are not going to that city from any other one.
             */
            sourceCityName = it->first;
            break;
        }
    }
    /*
     * Now that, We have found our start city, use the initial mapping to traverse the route till the END.
     */
    it = boardingPasses.find(sourceCityName);
    cout << "\nfollowing is the sorted order of the boarding passes for your trip.\n\n";
    while(it!=boardingPasses.end()) {
        cout << (it->first) << " " << (it->second) << endl;
        it = boardingPasses.find(it->second);
    }
}

int main() {
    string sourceCityName = "", destinationCityName = "";
    int numberOfPasses = 0;
    /*
     * boardingPasses map contains the information about the
     * passes in the form of a mapping : travel_from_city ---> travel_to_city
     */
    unordered_map<string, string> boardingPasses;
    cout << "Please provide the input in following format.\n\nN (number of passes)\nSource-City destination-City\n.\n.\n.\nN times\n\n";
    cin >> numberOfPasses;
    for (int i=1; i <= numberOfPasses; i++) {
        cin >> sourceCityName;
        cin >> destinationCityName;
        boardingPasses[sourceCityName] = destinationCityName;
    }
    sortTheBoardingPasses(boardingPasses);
    return 0;
}

File: QuesFromTest/01/python/sort-boarding-passes.py
'''
Name : Jitendra Kumar
Email: jitendra.iitp@gmail.com
'''

def sort_the_boarding_passes(boarding_passes):
    '''
    Print the boarding_passes in a manner such that,
    You start from your source_city and complete your journey
    through all the cities all the way till the destination_city
    '''
    reverse_map = dict() #reverse map will contains the mapping in reverse.

    for source_city in boarding_passes:
        '''
        reversing the initial mapping.
        Example: boarding_passes['bangalore'] = 'delhi' ===> reverse_map['delhi'] = 'bangalore'
        '''
        dest_city = boarding_passes[source_city]
        reverse_map[dest_city] = source_city

    source_city = None

    for source in boarding_passes:
        if not source in reverse_map:
            '''
            The city which does not have a 'travel_from_city' entry in reverse map,
            Actually is the starting point because you are not going to that city from any other one.
            '''
            source_city = source
            break
    '''
    Now that, We have found our start city,
    Use the initial mapping to traverse the route till the END.
    '''

    city = source_city
    while city:
        if (not city) or (not city in boarding_passes):
            break;
        print (city + " ---> " + boarding_passes[city])
        city = boarding_passes[city] or None

if __name__ == '__main__':
    try:
        message = "\nPlease provide the input in following format.\n\n"\
                "N (number of passes)\n"\
                "source_city destination_city\n"\
                ".\n"\
                ".\n"\
                ".\nN times\n\n"
        print (message)
        '''
        Using dict() for hashing the boarding_passes
        dict() lookup is O(1)
        '''
        number_of_passes = int(input())
        boarding_passes = dict()
        for x in range(number_of_passes):
            '''
            Example: delhi goa
            '''
            temp = input().split()
            source_city = temp[0]
            dest_city = temp[1]
            boarding_passes[source_city] = dest_city
        '''
        Sort the boarding_passes from start city to end.
        '''
        sort_the_boarding_passes(boarding_passes)
    except Exception as exc:
        print (exc)
